src/com/amazon/corretto/crypto/provider/AesGcmSpi.java

```diff
@@ -739,16 +775,31 @@ private int engineDecryptFinal(
                 // The cost of calling decryptAADBuf.getDataBuffer() when its buffer is empty is
                 // significant for 16-byte decrypt operations (approximately a 7% performance hit).
                 // To avoid this, we reuse the same empty array
-                decryptAADBuf.size() != 0 ? decryptAADBuf.getDataBuffer() : EMPTY_ARRAY,
+                decryptAADBuf.isEmpty() ? EMPTY_ARRAY : decryptAADBuf.getDataBuffer(),
                 decryptAADBuf.size());
-
-        if (ptrOut != null) {
-          context = new NativeContext(ptrOut[0]);
-        }
+        context = new NativeContext(ptrOut[0]);
+        return outlen;
       }
-      // Decryption completed successfully.
-      return outLen;
-    } catch (AEADBadTagException e) {
+      // We don't have a context, and we don't need to save it
+      return oneShotDecrypt(
+          0,
+          false,
+          null,
+          workingInputArray,
+          workingInputOffset,
+          workingInputLength,
+          output,
+          outputOffset,
+          tagLength,
+          key,
+          iv,
+
+          // The cost of calling decryptAADBuf.getDataBuffer() when its buffer is empty is
+          // significant for 16-byte decrypt operations (approximately a 7% performance hit).
+          // To avoid this, we reuse the same empty array
+          decryptAADBuf.size() != 0 ? decryptAADBuf.getDataBuffer() : EMPTY_ARRAY,
```

Comment:
nit: why use `.size()` here and `.isEmpty()` on L778 and L754? is there a difference?

Reply:
There is no difference. I rebased and this one was not updated. I'll change it.